src/inventory_ui.cpp

```diff
@@ -108,308 +393,413 @@ bool inventory_column::is_selected( const inventory_entry &entry ) const
+                } else if( entry.chosen_count >= entry.get_available_count() ) {
+                    mvwputch( win, yy, xx, c_ltgreen, '+' );
+                } else {
+                    mvwputch( win, yy, xx, c_ltgreen, '#' );
+                }
+            }
         }
     }
 }
 
-void selection_column::prepare_paging( size_t new_entries_per_page )
-{
-    inventory_column::prepare_paging( new_entries_per_page );
-    if( entries.empty() ) { // Category must always persist
-        entries.emplace_back( selected_cat.get() );
-    }
-}
-
 void selection_column::on_change( const inventory_entry &entry )
 {
     inventory_entry my_entry( entry, selected_cat.get() );
 
     const auto iter = std::find( entries.begin(), entries.end(), my_entry );
-    if( my_entry.chosen_count != 0 ) {
-        if( iter == entries.end() ) {
+    if( iter == entries.end() ) {
+        if( my_entry.chosen_count != 0 ) {
             add_entry( my_entry );
-        } else {
-            iter->chosen_count = my_entry.chosen_count;
         }
+    } else if( my_entry.chosen_count == 0 ) {
+        remove_entry( my_entry );
     } else {
-        entries.erase( iter );
+        iter->chosen_count = my_entry.chosen_count;
+        expand_to_fit( my_entry );
     }
 
     prepare_paging();
     // Now let's update selection
     const auto select_iter = std::find( entries.begin(), entries.end(), my_entry );
     if( select_iter != entries.end() ) {
-        select( std::distance( entries.begin(), select_iter ) );
+        select( std::distance( entries.begin(), select_iter ), 1 );
     } else {
-        select( entries.empty() ? 0 : entries.size() - 1 ); // Just select the last one
+        select( entries.empty() ? 0 : entries.size() - 1, -1 ); // Just select the last one
     }
 }
 
-std::string selection_column::get_entry_text( const inventory_entry &entry ) const
+inventory_input inventory_selector::get_input()
 {
-    std::ostringstream res;
-
-    if( entry.is_item() ) {
-        if( OPTIONS["ITEM_SYMBOLS"] ) {
-            res << entry.get_item().symbol() << ' ';
-        }
+    inventory_input res;
 
-        const size_t available_count = entry.get_available_count();
-        if( entry.chosen_count > 0 && entry.chosen_count < available_count ) {
-            res << string_format( _( "%d of %d"), entry.chosen_count, available_count ) << ' ';
-        } else if( available_count != 1 ) {
-            res << available_count << ' ';
-        }
-        res << entry.get_item().display_name( available_count );
-    } else {
-        res << inventory_column::get_entry_text( entry ) << ' ';
+    res.action = ctxt.handle_input();
+    res.ch = ctxt.get_raw_input().get_first_input();
+    res.entry = find_entry_by_invlet( res.ch );
 
-        if( entries.size() > 1 ) {
-            res << string_format( "(%d)", entries.size() - 1 );
-        } else {
-            res << _( "(NONE)" );
-        }
+    if( res.entry != nullptr && !res.entry->is_selectable() ) {
```

Comment:
Overly verbose - that's one logical operation:

`return res && res.entry->is_selectable() ? res : nullptr`